src/attr/_make.py

```diff
@@ -871,22 +882,31 @@ def __init__(self, name, default, validator, repr, cmp, hash, init,
         bound_setattr("convert", convert)
         bound_setattr("metadata", (metadata_proxy(metadata) if metadata
                                    else _empty_metadata_singleton))
+        bound_setattr("type", type)
 
     def __setattr__(self, name, value):
         raise FrozenInstanceError()
 
     @classmethod
-    def from_counting_attr(cls, name, ca):
+    def from_counting_attr(cls, name, ca, type=None):
+        # type holds the annotated value. deal with conflicts:
+        if type is None:
+            type = ca.type
+        elif ca.type is not None and type is not ca.type:
+            raise ValueError(
```

Comment:
I think we should raise an error regardless whether they differ?  A tighter API envelope is usually better.